### PR TITLE
New variables structure

### PR DESCRIPTION
**Pull request by noseglid**
Do not include storage modifiers in the variables matches
because it's not great to re-use them in code blocks in that
case.

This also made it possible to put `#variables` as an
include in `#code`, which removes the need for having
both `#code` and `#variables` in many blocks.

One difference here is that storage modifiers (`private`, `final`,
`static`, etc) will no longer be tagged with `meta.definition.variable.java`,
and it shouldn't be IMO. It's not part of the definition, but rather a modification to
a definition. Syntax highlighting remains the same.

This also fixes my comment in #47, but not #47 itself.

Fixes: #50

### CODE REVIEW

File: grammars/java.cson [hunk 1]
```diff
@@ -377,19 +391,7 @@
...
-        'end': '}'
-        'endCaptures':
-          '0':
-            'name': 'punctuation.section.block.end.bracket.curly.java'
-        'patterns': [
-          {
-            'include': '#code'
-          }
-        ]
+        'include': '#anonymous-block-and-instance-initializer'
```
Comment (noseglid) [inline grammars/java.cson]: No actual changes, just some splitting up code to make it more readable.

File: grammars/java.cson [hunk 2]
```diff
@@ -999,50 +1001,54 @@
       }
     ]
   'variables':
-    'applyEndPatternLast': 1
+    'begin': '''
+      (?x:(?=
+        (?:
```
Comment (50Wliu) [inline grammars/java.cson]: Since we're in a lookahead all the non-capturing groups aren't needed.

File: grammars/java.cson [hunk 3]
```diff
@@ -999,50 +1001,54 @@
...
-              |
-              (?:(?:[a-z]\\w*\\s*\\.\\s*)*[A-Z]+\\w*) # object type
-            )
-            \\s+
-            (?!private|protected|public|native|synchronized|abstract|threadsafe|transient|static|final)
-            [\\w\\d_<>\\[\\],\\?][\\w\\d_<>\\[\\],\\?\\s]*
-            \\s*(=|;)
-          ))
-        '''
+        'match': '([A-Za-z$_][\\w$]*)(?=(\\[\\])?\\s*(?=;|=|,))'
```
Comment (50Wliu) [inline grammars/java.cson]: There can be spaces before the array brackets, eg `int test [] = new int[3];`.

File: grammars/java.cson [hunk 4]
```diff
@@ -999,50 +1001,54 @@
...
+        'captures':
+          '1':
+            'name': 'variable.definition.java'
```
Comment (50Wliu) [inline grammars/java.cson]: I realized after I merged the PR adding this that the scope should probably be `variable.other.definition.java` instead.

File: grammars/java.cson [hunk 5]
```diff
@@ -999,50 +1001,54 @@
...
-              }
-            ]
-          }
-          {
             'include': '#code'
           }
         ]
       }
+      {
+        'include': '#code'
```
Comment (50Wliu) [inline grammars/java.cson]: Why is this needed?

File: grammars/java.cson [hunk 6]
```diff
@@ -999,50 +1001,54 @@
...
+      }
     ]
+    'end': '(?=;)'
+    'name': 'meta.definition.variable.java'
+  'member-variables':
+      'begin': '(?=private|protected|public|native|synchronized|abstract|threadsafe|transient|static|final)'
```
Comment (50Wliu) [inline grammars/java.cson]: Your indentation is off here

File: grammars/java.cson [hunk 7]
```diff
@@ -999,50 +1001,54 @@
...
             'include': '#code'
           }
         ]
       }
+      {
+        'include': '#code'
+      }
     ]
+    'end': '(?=;)'
+    'name': 'meta.definition.variable.java'
```
Comment (50Wliu) [inline grammars/java.cson]: As in the other PR, end and name before patterns.

File: spec/java-spec.coffee [hunk 8]
```diff
@@ -60,9 +60,10 @@ describe 'Java grammar', ->
   it 'tokenizes punctuation', ->
     {tokens} = grammar.tokenizeLine 'int a, b, c;'
 
-    expect(tokens[2]).toEqual value: ',', scopes: ['source.java', 'punctuation.separator.delimiter.java']
-    expect(tokens[4]).toEqual value: ',', scopes: ['source.java', 'punctuation.separator.delimiter.java']
-    expect(tokens[6]).toEqual value: ';', scopes: ['source.java', 'punctuation.terminator.java']
+    console.log tokens
```
Comment (50Wliu) [inline spec/java-spec.coffee]: :scream:

File: grammars/java.cson (hunk 2 above)
Comment (noseglid) [inline grammars/java.cson]: Right you are. Copied this from the previous definition, but might as well simplify!

File: grammars/java.cson (hunk 3 above)
Comment (noseglid) [inline grammars/java.cson]: 👍

File: grammars/java.cson (hunk 4 above)
Comment (noseglid) [inline grammars/java.cson]: Yes, that seems to be the consesus in textmate. Changing.

File: grammars/java.cson (hunk 5 above)
Comment (noseglid) [inline grammars/java.cson]: To get e.g. punctuation in `int a, b, c;`

File: grammars/java.cson (hunk 7 above)
Comment (noseglid) [inline grammars/java.cson]: 👍

File: grammars/java.cson (hunk 6 above)
Comment (noseglid) [inline grammars/java.cson]: Nice catch. Seems to work all the same. I hate CoffeeScript/cson for this reason. "Indentation is dead important! sometimes..."

File: spec/java-spec.coffee (hunk 8 above)
Comment (noseglid) [inline spec/java-spec.coffee]: ouch. sloppy :( I blame late night coding

File: grammars/java.cson [hunk 9]
```diff
@@ -999,50 +1001,54 @@
       }
     ]
   'variables':
-    'applyEndPatternLast': 1
```
Comment (noseglid) [inline grammars/java.cson]: I removed this.  To be frank, I don't know how It worked with it before, since `;` is matched by `#code` which was included before too.

Comment (50Wliu) [inline grammars/java.cson]: Surprisingly first-mate actually recognizes this.  The super-sparse documentation seems to suggest that this only matters when the end patterns of a subpattern and the overarching `end` are exactly the same, but if everything already works, then no need to have it!

Comment (noseglid) [inline grammars/java.cson]: Cool. In my manual tests it seems to work well, and the specs agrees.

**Comment by noseglid** [comment]
NB. Added this to commit message
> One difference here is that storage modifiers (`private`, `final`,
`static`, etc) will no longer be tagged with `meta.definition.variable.java`,
which is correct IMO. It's not a definition, but rather a modification to
a definition. Syntax highlighting remains the same.

**Comment by 50Wliu** [comment]
:boat:

**Comment by 50Wliu** [comment]
A multi-array variable still isn't highlighted correctly, eg `private int stats[][];`.

**Comment by noseglid** [comment]
screen shot 2016-06-13 at 9 01 35 am

Looks okay to me... what do you mean?
